Fix issue with trending order when user has chosen languages

mjankowski opened the pull request:
Resolves https://github.com/mastodon/mastodon/issues/33555

The specs added in https://github.com/mastodon/mastodon/pull/33531 only covered the base "default locale" case, not the scenario where the account querying for trends has chosen languages set. The locale comes in as a flat single value, whereas the languages come in as an array. With languages set, we effectively had a double array which the query builder could not quote (hence `TypeError` in linked issue).

I believe this change resolves this issue in the sense of there no longer being errors, and both the "default locale" and "chosen languages" scenarios being handled correctly (without runtime error) ... but I also want to do some followup here to contemplate whether we also had an unintended behavior change around how the languages are ordered/prioritized relative to each other.

mjankowski committed:
Fix issue with trending order when user has chosen languages

mjankowski commented:
More followup...

Up until https://github.com/mastodon/mastodon/pull/33531 the logic was like:

- When a user does NOT have chosen languages rank the supplied locale first, then order by score
- When a user DOES have chosen languages, treat "any of those languages" equally (similar to locale in first case), then order by score

That change failed to cover the user languages case, and had code-level breakage described by the issue. This PR does fix that code breakage, in the sense that the specific error is handled here.

However, there's another unintended side effect, which is that the logic for when a user has chosen languages effectively changes to rank the languages, instead of treating "any of these languages" equally.

So, this PR is an improvement in that it fixes the error, and we could work on logic fix as followup.

Alternately, I can do a PR to revert the code-level changes, but keep the spec-level changes from original PR and this one (they will both assist in any future re-attempt at this), and add a spec that would fail because of this unintended logic change as well.

ClearlyClaire commented:
> Alternately, I can do a PR to revert the code-level changes, but keep the spec-level changes from original PR and this one (they will both assist in any future re-attempt at this), and add a spec that would fail because of this unintended logic change as well.

I would be in favor of that, personally.

mjankowski commented:
Updated to a revert of `in_order_of`.

Net result of this and the original PR would be:

- Added specs for the base "default locale" situation when calling records
- Added specs for when the user has chosen languages, with enough detail that the ordering interaction between languages/scores is respected, to catch any future refactor
- Restore the arel nodes case ordering this, but do a tiny DRY refactor to move that call into the base class
